[GraphQL] Permissions

carlwoodhouse opened the pull request:
This pr adds the ability for more flexible/granular permissions in graphql.

Currently i've opted to keep a single schema (rather then permission based schema) as it would be a heavy perf hit to have to cache (or not cache) the schema. Reading online the choice in graphql seems to be with us. So fields you dont have permission to view the content for are still discoverable via the schema.

**Done**
* Added new permission `ExecuteGraphQLMutations` this is required in order to execute any mutations regardless of type specific permissions. I didnt do anything else for mutations as they aren't really fully wired up anyway (no default ones exist).
* `ExecuteGraphQL` permission is still required to execute any type of permission
* Added permission validation rule (https://graphql-dotnet.github.io/docs/getting-started/query-validation)
* Added `RequirePermission` extension for fields/fieldbuilders, so we can make any field (and because of how graphql this also means types) require specific permissions. This extension will be usable by anyone on customgraphtypes :) 
* By default i've wired up the `View {contentype}` permission for each contenttype in the contenttypefieldbuilder, meaning people can exclude types using the existing role/permission system from being viewed via query execution.
* Added the ability to use permission context in graphql query tests; to test permissions

~~I've also changed the default error handling, currently if any execution error occurs our graphql implementation throws a 400 error, as far as i can tell this isn't actually correct [(though again this seems to be down to interpretation)](https://itnext.io/the-definitive-guide-to-handling-graphql-errors-e0c58b52b5e1) - we should be throwing 200 with the error's property correctly populated so we can give partial results. Eg. if someone queried multiple types, one they had permission to view and one they didnt, they'd recieve the data for the one they are allowed to view and error info on the the one they are not.~~ This seems to be too down to interpretation, i've reverted to returning 400 if errors are thrown, unless they are auth errors then its returning 401; this mirrors the behaviour of other popular servers such as graphql-dotnet-server, graphql-express and apollo's graphql server.

Jetski5822 commented:
Really like it - code is nice man

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs:
```diff
@@ -165,13 +165,10 @@ private async Task ExecuteAsync(HttpContext context, ISchemaFactory schemaServic
                 _.Inputs = request.Variables.ToInputs();
                 _.UserContext = _settings.BuildUserContext?.Invoke(context);
                 _.ExposeExceptions = _settings.ExposeExceptions;
+                _.ValidationRules = (_settings.ValidationRules ?? Enumerable.Empty<IValidationRule>()).Concat(DocumentValidator.CoreRules());
```
no need to do ?? here... _settings.ValidationRules is a list and will never be null.

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs:
```diff
@@ -165,13 +165,10 @@ private async Task ExecuteAsync(HttpContext context, ISchemaFactory schemaServic
-                : HttpStatusCode.OK;
-
-            context.Response.StatusCode = (int)httpResult;
+            context.Response.StatusCode = (int) HttpStatusCode.OK;
```
why?

Jetski5822 replied:
Surely with a permissions check now as well you need to return the http status code for forbidden?

carlwoodhouse replied:
@Jetski5822 did you read the comment about it in the PR description ?

Jetski5822 replied:
nope :P

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs:
```diff
@@ -0,0 +1,72 @@
+           
+                    if (op.OperationType == OperationType.Mutation)
+                    {
+                        var authorizationManager = context.ServiceProvider.GetService<IAuthorizationService>();
```
context could be null

Jetski5822 replied:
I would hard case it (GraphQLContext)validationContext.UserContext on line 16

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs:
```diff
@@ -0,0 +1,72 @@
+            {
+                _.Match<Operation>(op =>
+                {
+           
```
remove dead line

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs:
```diff
@@ -0,0 +1,72 @@
+                        validationContext.ReportError(new ValidationError(
+                            validationContext.OriginalQuery,
+                            "Forbidden",
+                            $"Authorization is required to access this field.",
```
remove $

Jetski5822 replied:
or provide field name

Jetski5822 replied:
IStringLocalizer

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs:
```diff
@@ -0,0 +1,72 @@
+                        validationContext.ReportError(new ValidationError(
+                            validationContext.OriginalQuery,
+                            "Forbidden",
+                            $"Authorization is required to access {op.Name}.",
```
localization

Jetski5822 commented on src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/PermissionsExtensions.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+            var permissions = type.GetMetadata<List<Tuple<Permission, object>>>(PermissionsKey);
+
+            if (permissions is null)
```
can this ever be null? its an enumerable so I dont think this check is needed

Jetski5822 replied:
Side node ( is null = nice syntax )

Skrypt replied:
Should you not just use `if (permissions == null)` like @sebastienros asked?

carlwoodhouse replied:
yes it can be null, metadata is just a dictionary keyed by a key chosen by us, we need to add it if it doesn't exist.

Jetski5822 commented on src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/PermissionsExtensions.cs:
```diff
@@ -0,0 +1,32 @@
+
+            if (permissions is null)
+            {
+                type.Metadata[PermissionsKey] = permissions = new List<Tuple<Permission, object>>();
```
is above is true, no need for this check

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs:
```diff
@@ -184,8 +181,10 @@ private async Task WriteErrorAsync(HttpContext context, string message, Exceptio
-            errorResult.Errors = new ExecutionErrors();
+            var errorResult = new ExecutionResult
+            {
+                Errors = new ExecutionErrors()
```
hmmm - do we never return errors?

carlwoodhouse replied:
yes they are added after, in different ways depending if an exception was passed to this method.

carlwoodhouse commented:
Will change; habit :)

On Thu, 25 Jul 2019 at 04:54, Jasmin Savard <notifications@github.com>
wrote:

> *@Skrypt* commented on this pull request.
> ------------------------------
>
> In
> src/OrchardCore/OrchardCore.Apis.GraphQL.Abstractions/Extensions/PermissionsExtensions.cs
> <https://github.com/OrchardCMS/OrchardCore/pull/3951#discussion_r307105392>
> :
>
> > +using System.Collections.Generic;
> +using GraphQL.Builders;
> +using GraphQL.Types;
> +using OrchardCore.Security.Permissions;
> +
> +namespace OrchardCore.Apis.GraphQL
> +{
> +    public static class PermissionsExtensions
> +    {
> +        public static readonly string PermissionsKey = "Permissions";
> +
> +        public static void RequirePermission(this IProvideMetadata type, Permission permission, object resource = null)
> +        {
> +            var permissions = type.GetMetadata<List<Tuple<Permission, object>>>(PermissionsKey);
> +
> +            if (permissions is null)
>
> Should you not just use if (permissions == null) like @sebastienros
> <https://github.com/sebastienros> asked?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/OrchardCore/pull/3951?email_source=notifications&email_token=AADRUEGA53ZJ2ULJM7HQSSDQBEPYFA5CNFSM4IE5OZOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7QMKYY#discussion_r307105392>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADRUECNB3HN63DNIUL36UDQBEPYFANCNFSM4IE5OZOA>
> .
>

carlwoodhouse commented:
Reverted always returning 200; this seems to be too down to interpretation, i've reverted to returning 400 if errors are thrown, unless they are auth errors then its returning 401; this mirrors the behaviour of other popular servers such as graphql-dotnet-server, graphql-express and apollo's graphql server.

carlwoodhouse commented:
Ok, mostly done ..

I worked out how to to pass permission context for the tests; though honestly there might be a better way of doing it; open to suggestions

carlwoodhouse commented:
Added more tests, tidied some stuff up.

Is there anything anybody thinks needs doing this to finish it? would be nice to get it merged as permissions is quite important; nice to get the validation extension point in too ...

/cc @sebastienros

jrestall commented:
Nice work! This will be really useful. I've reviewed the changes and don't see any issues, I like that you are using the existing ViewContent permission. This will make GraphQL permissions a lot more granular and flexible.

LGTM :shipit:

agriffard commented:
Conflicts to resolve before merging.

agriffard commented on test/OrchardCore.Tests/Apis/GraphQL/Queries/RecentBlogPostsQueryTests.cs:
```diff
@@ -5,7 +5,7 @@
 using Xunit;
 
-namespace OrchardCore.Tests.Apis.GraphQL.Queries
+namespace OrchardCore.Tests.Apis.GraphQL
```
Why do you change the namespace?

carlwoodhouse replied:
for better top level grouping of the tests in explorer ;)

eg ..

![image](https://user-images.githubusercontent.com/465424/69811427-f2fbe980-11e5-11ea-95e9-9f8f4e2063a4.png)

carlwoodhouse commented:
have an issue with a failing test since resolving conflicts, investigating it

jrestall commented on src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs:
```diff
@@ -0,0 +1,67 @@
+
+            // awaitable IValidationRule in graphql dotnet is coming soon:
+            // https://github.com/graphql-dotnet/graphql-dotnet/issues/1140
+            return type.GetPermissions().All(x => authorizationManager.AuthorizeAsync(context.User, x.Permission, x.Resource).GetAwaiter().GetResult());
```
Looks like support for ValidateAsync has landed. Do you want to update as part of this PR or another?
https://github.com/graphql-dotnet/graphql-dotnet/pull/1371

carlwoodhouse commented:
Probably should be another pr; as the new version of graphql dotnet has
some breaking changes round the user context we’ll need to sort out

On Thu, 28 Nov 2019 at 20:15, James Restall <notifications@github.com>
wrote:

> *@jrestall* approved this pull request.
>
> All looks great, be good to use ValidateAsync but could be done as a new
> PR. graphql-dotnet/graphql-dotnet#1371
> <https://github.com/graphql-dotnet/graphql-dotnet/pull/1371>
> ------------------------------
>
> In
> src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/ValidationRules/RequiresPermissionValidationRule.cs
> <https://github.com/OrchardCMS/OrchardCore/pull/3951#discussion_r351918430>
> :
>
> > +                            validationContext.OriginalQuery,
> +                            ErrorCode,
> +                            localizer["Authorization is required to access the field. {0}", fieldAst.Name],
> +                            fieldAst));
> +                    }
> +                });
> +            });
> +        }
> +
> +        private static bool Authorize(IProvideMetadata type, GraphQLContext context)
> +        {
> +            var authorizationManager = context.ServiceProvider.GetService<IAuthorizationService>();
> +
> +            // awaitable IValidationRule in graphql dotnet is coming soon:
> +            // https://github.com/graphql-dotnet/graphql-dotnet/issues/1140
> +            return type.GetPermissions().All(x => authorizationManager.AuthorizeAsync(context.User, x.Permission, x.Resource).GetAwaiter().GetResult());
>
> Looks like support for ValidateAsync has landed. Do you want to update as
> part of this PR or another?
> graphql-dotnet/graphql-dotnet#1371
> <https://github.com/graphql-dotnet/graphql-dotnet/pull/1371>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrchardCMS/OrchardCore/pull/3951?email_source=notifications&email_token=AADRUEBFMG3R6FEUG3LGDY3QWAREXA5CNFSM4IE5OZOKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCNLKZOY#pullrequestreview-324447419>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADRUECSMX4H7X4MIYJ467TQWAREXANCNFSM4IE5OZOA>
> .
>
